Feature - 31 - GC Notify Integration

esizer opened the pull request:
🤖 Resolves #31

## 👋 Introduction

This presents a solution to send emails via [GC Notify](https://notification.canada.ca/)

## 🕵️ Details

This binds a custom GC Notify Client (`App\Notify\Client`) to Laravel's  [Service Container](https://laravel.com/docs/9.x/container) using a [Service Provider](https://laravel.com/docs/9.x/providers) (`App\Providers\NotifyServiceProvider`) and [Facade](https://laravel.com/docs/9.x/facades) (`App\Facades\Notify`).

### Usage

This only implements a basic solution and a test api routes to show an example of usage. We expose the Notify client as a singleton to be interacted with directly.

```php
<?php

use App\Facades\Notify;

// Note: $repsonse is an instance of \Illuminate\Http\Client\Response

// Email Attachment
$attachment = [
  'file' => $file, // Base64 encoded file
  'filename' => 'file.txt' // Name of the file (can be anything)
];

// Sending email
$response = Notify::sendEmail(
  $to, // Email of receipient
  $templateId, // This is an ID of a template created in GC Notify
  $personalisation, // (optional) An array of template values to be replaced by
  $reference, // (optional) Add a reference key to identify the message
  $replyTo, // (optional) ID for a reply to email address
  $attachment, // (optional) A file attachment for the email
);

// Sending SMS
$response = Notify::sendSms(
  $phoneNumber, // Phone number of receipient
  $templateId, // This is an ID of a template created in GC Notify
  $personalisation, // An array of template values to be replaced by
  $reference, // (optional) Add a reference key to identify the message
  $sender // (optional) ID for a sender number
);

// Bulk Recipient Rows (email)
$rows = [
  ['email address', 'name'],
  ['email+1@domain.com', 'First User'],
  ['email+2@domain.com', 'Second User']
];

// Bulk Recipient Rows (phone)
$rows = [
  ['phone number', 'name'],
  ['+5555555555', 'First User'],
  ['+4444444444', 'Second User']
];

// Sending Bulk Email/SMS
$response = Notify::sendBulk(
  $name, // Unique Indentifier for bulk messages
  $rows, // Array of recipients, first index is headers, following is data
  $templateId, // This is an ID of a template created in GC Notify
  $scheduleFor, // (optional) \Carbon\Carbon date to schedule when messages send
  $replyTo, // (optional) ID for a reply to email address
);
```

### Configuration

You can find configuration in the `config/notify.php` file. It is required to have an environment variable to be set for `GCNOTIFY_API_KEY`. You can get this from @esizer.

https://github.com/GCTC-NTGC/gc-digital-talent/blob/da4dc10b4f85c786e974ff264d62f71268a1cd21/api/config/notify.php#L17-L32

## 🧪 Testing

> **Note**
> You may need to shell into the container and run `php artisan config:cache` to get the new notify config.

1. Contact @esizer on Slack and ask for your email or phone number to be added to the safe list (we are allowed 5 of each in testing)
2. @esizer will respond with an API key that you will add to your `api/.env` file as `GCNOTIFY_API_KEY="provided-key"`
3. Refresh the API `docker-compose run --rm maintenance bash refresh_api.sh`
4. Open to `app/routes/api.php` replace the commented areas with your email address/phone number you sent me
5. Navigate to the routes to test `http://localhost:8000/api/*` (see list of endpoints following instructions)
5. You should see the json response from the API
7. Wait for the email to enter your inbox

### Endpoints

* /notify/email: Sending Emails
* /notify/sms: Sending SMS
* /notify/email/attachment: Send email with attachment (requires adding file in specified location)
* /notify/bulk/email: Send bulk emails
* /notify/bulk/sms: Send bulk SMS

## 📸 Screenshot

### Email

<img width="596" alt="Screenshot 2023-01-16 at 2 25 36 PM" src="https://user-images.githubusercontent.com/4127998/212752978-cebcc4b6-f656-4396-8046-7918b1356c89.png">

### SMS

<img width="460" alt="Screenshot 2023-01-17 at 1 25 47 PM" src="https://user-images.githubusercontent.com/4127998/212981050-495ec257-02b4-40d9-8507-37d0a7502b72.png">

mnigh commented on api/_lighthouse_ide_helper.php:
```diff
@@ -155,12 +155,3 @@ public function assertGraphQLValidationPasses(): self
         }
     }
 }
-
```
> This file should be ignored by git as it can be autogenerated.

Is this comment still relevant or should this file be added to `.gitignore`?

esizer replied:
I can't be certain but I think it should be safe to add it to `.gitignore`

tristan-orourke replied:
Are we checking this in to make some CI easier?

esizer replied:
That is a very good question. I think you may be correct but I am not sure who would be able to verify this 🤷‍♀️

mnigh commented on api/app/Providers/GCNotifyServiceProvider.php:
```diff
@@ -0,0 +1,38 @@
+                'httpClient' => new GuzzleClient
+            ]);
+
+            return new Client($config);
```
Undefined type 'Alphagov\Notifications\Client'

mnigh commented on api/routes/api.php:
```diff
@@ -18,3 +18,17 @@
 
+Route::get('/test-notify', function () {
+    try {
+        Notify::sendEmail(
```
Undefined type 'Notify'

esizer replied:
This is because aliases are resolved at runtime. I don't think we really need to use it as an alias so I updated it in [fix facade undefined](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/4e9d54453c406550ccdc28d722de88b6ebd02c73)

mnigh commented on api/schema-directives.graphql:
```diff
@@ -202,7 +202,7 @@ directive @clearCache(
   """
   field: String
-) on FIELD_DEFINITION
+) repeatable on FIELD_DEFINITION
```
Any reason why the schema directives needed to change for this branch?

esizer replied:
I have no idea what this was all about 🤷

mnigh commented on api/routes/api.php:
```diff
@@ -18,3 +18,17 @@
+    try {
+        Notify::sendEmail(
+            'email@domain.com', // Replace with the email you gave to @esizer
+            'c386053b-e65c-477e-8f09-da19c2e9dea0',
```
I know this is just an experimental branch, but the `templateId` would probably be best as an environment variable as well, oui?

esizer replied:
So, the way it works, we would need to add a lot of template id's and since they aren't really secret and need to scale I don't think we need to be putting them in as environment variables. What we could do is maybe make a map like we did `ApiEnums` so it is more readable. i.e

```php
<?php

namespace App\Facades;

class NotifyEnums
{
    const STATUS_UPDATE_TEMPLATE = 'template-id';

    public static function templates(): array
    {
        return [
            self::STATUS_UPDATE_TEMPLATE
        ];
    }
}

# Usage
$templates = NotifyEnums::templates();
Notify::sendEmail(
    'email@domain.com',
    NotifyEnums::STATUS_UPDATE_TEMPLATE,
    []
);
```

mnigh replied:
I like that, much easier to read than an id.

mnigh commented on api/composer.json:
```diff
@@ -6,6 +6,7 @@
     "license": "AGPL-3.0",
     "require": {
         "php": "^8.1",
+        "alphagov/notifications-php-client": "^5.0",
```
I know this is part of GC Notify's official [documentation](https://documentation.notification.canada.ca/en/clients.html) for client libraries, but are any of these features ones we will require that are not available using the UK's PHP client?

> Feature differences to keep in mind
> - [Sending files by email is different with GC Notify](https://documentation.notification.canada.ca/en/send.html#sending-files-by-email-is-an-API-only-feature)
> - Sending bulk notifications through the GOV.UK Notify clients is not directly supported. You can still use one if the client supports customization of the URL endpoint
> - Sending letters is not available
> - Receiving text messages is not available

esizer replied:
Yeah, I'm willing to take this further and write a custom client so we can implement the missing features. I don't think it would be super complex and since this was so quick to implement we still have plenty of time in the 3 day timebox.

@tristan-orourke What are your thoughts on this?

mnigh reviewed (commented):
This is very cool, nice work! 🌟 

A few linting notes and some questions that might be out of the scope of this experiment but I believe worthy of consideration for eventual implementation.

tristan-orourke commented:
> I'm willing to take this further and write a custom client so we can implement the missing features. I don't think it would be super complex and since this was so quick to implement we still have plenty of time in the 3 day timebox.

If I understand correctly, you mean essentially writing a PHP class which builds the payloads/requests to send directly to the [GC Notify api](https://documentation.notification.canada.ca/en/send.html#sending-notifications-in-bulk), so we can send files or messages in bulk?

That sounds very doable and very useful. I think go for it! 👍

esizer commented:
> If I understand correctly, you mean essentially writing a PHP class which builds the payloads/requests to send directly to the [GC Notify api](https://documentation.notification.canada.ca/en/send.html#sending-notifications-in-bulk), so we can send files or messages in bulk?
> 
> That sounds very doable and very useful. I think go for it! 👍

Yup! I think it makes sense here since as @mnigh pointed out, the Gov UK one does not support all the features of GC Notify and I see some of these features as very useful particularly sending bulk notifications.

I'll start work on it 👍

esizer commented:
@tristan-orourke and @mnigh I have updated this to use a custom client so we can add on functionality as needed and cover functionality not covered in the UK Gov one. Right now there are just three functions available:

 - Send Email
 - Send SMS
 - Send Bulk Email/SMS

## Templates

For readability, I moved the template IDs to `app/config/notify.php` so when we access them it is a little more readable.

### Config

https://github.com/GCTC-NTGC/gc-digital-talent/blob/da4dc10b4f85c786e974ff264d62f71268a1cd21/api/config/notify.php#L17-L32

### Usage

```php
// Access a single template directly
$template = config('notify.templates.email');

// Get all templates as array
$templates = config('notify.templates');
$emailTemplates = $templates['email'];
```
## Example Usage

You can see some example usage in the `api/routes/api.php` file.

> **Note**
> These will be removed before merging

https://github.com/GCTC-NTGC/gc-digital-talent/blob/da4dc10b4f85c786e974ff264d62f71268a1cd21/api/routes/api.php#L30-L40

### Events and Listeners

The idea is that we would setup an [Event](https://laravel.com/docs/9.x/events#defining-events) and a [Listener](https://laravel.com/docs/9.x/events#defining-listeners) similar to what we do for the application snapshot.

So, We could have an event `ApplicationStatusChanged` and `SendApplicationStatusEmail` the when an admin updates a users status we just dispatch the event and a notification email should send.

#### Application Submission Example

https://github.com/GCTC-NTGC/gc-digital-talent/blob/da4dc10b4f85c786e974ff264d62f71268a1cd21/api/app/Events/ApplicationSubmitted.php

https://github.com/GCTC-NTGC/gc-digital-talent/blob/da4dc10b4f85c786e974ff264d62f71268a1cd21/api/app/Listeners/StoreApplicationSnapshot.php

https://github.com/GCTC-NTGC/gc-digital-talent/blob/da4dc10b4f85c786e974ff264d62f71268a1cd21/api/app/GraphQL/Mutations/SubmitApplication.php#L43

esizer commented on .vscode/project-words.txt:
```diff
@@ -44,6 +44,7 @@ nuage
 Nuwave
 OCIO
 Ojibway
+personalisation
```
This is an accepted spelling on personalization and comes from `en-GB` which is preferred in GC Notify over the `en-US` spelling.

mnigh commented on api/app/Notify/Client.php:
```diff
@@ -0,0 +1,282 @@
+
+        if(!is_null($scheduleFor)) {
+            if(!$scheduleFor->isFuture()) {
+                throw new NotFutureDateException("Schedule for must be a date in the future.");
```
Maybe "Schedule for date must be a date in the future." instead of "Schedule for must be a date in the future."?

esizer replied:
Yeah, that sounds better. Thanks!

[improve schedule date exception message](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/70da4f12fc1b845c9104f7b58f961ec8d9592152)

esizer commented on api/routes/api.php:
```diff
@@ -18,3 +20,79 @@
+    Route::get('email/attachment', function () use ($templates) {
+        try {
+            $file = Storage::get('example.txt'); // Add a file by this name to `api/storage/app`
+            $response = Notify::sendEmail(
+                'domain@test.com', // Replace with your email
+                $templates['email'],
+                ['name' => 'Your Name'],
+                null,
+                null,
+                [
+                    'file' => base64_encode($file),
+                    'filename' => "attachment.txt"
+                ]
+            );
```
I don't like the double null's to add an attachment. Is anyone aware of a better way to do this or should be shuffle the order around to put `$attachment` ahead of `$reference` and `$replyTo`?

mnigh replied:
I don't mind the double nulls, perhaps add a comment beside them?

mnigh commented on api/routes/api.php:
```diff
@@ -18,3 +20,79 @@
+                $response = Notify::sendBulk(
+                    'Test',
+                    [
+                        ['email address', 'name'],
```
I found the bulk stuff a bit confusing, not sure of a better solution though. What about separating the concerns for email and sms for bulk?

esizer replied:
Yeah, I agree on this. I was struggling to figure out how to do this. Maybe an associative array?

```php
  Notify::sendBulkEmail(
    'Test',
    [
       [ 
          'email' => 'email@domain',
          'personalisation' => ['name' => 'Test']
       ]
    ]
  );
  
    Notify::sendBulkSms(
    'Test',
    [
       [ 
          'phone_number' => '5555555555',
          'personalisation' => ['name' => 'Test']
       ]
    ]
  );
```

tristan-orourke replied:
I think that having two methods is definitely an improvement.

esizer replied:
okii okii, I have updated to split the bulk sms and email as well as changing the shape of the `$rows` arg to make it a little more readable....hopefully 🤞 

[update bulk notify API to split sms and email](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/850584e945c34cf4387df9a7acfa5b74c26dbc33)

See the API in the `NotifyTest`

https://github.com/GCTC-NTGC/gc-digital-talent/blob/850584e945c34cf4387df9a7acfa5b74c26dbc33/api/tests/Unit/NotifyTest.php#L66-L90

https://github.com/GCTC-NTGC/gc-digital-talent/blob/850584e945c34cf4387df9a7acfa5b74c26dbc33/api/tests/Unit/NotifyTest.php#L103-L127

tristan-orourke replied:
Nice! I would just suggest adding some details in the docstrings about the shape of `$rows`.

esizer replied:
Great idea, added in [cleanup + improve docs](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/2e8b6ca832c0d8ad2d923e837fece269b7f02663). Hopefully it is clear enough 😅

mnigh reviewed (commented):
Awesome work! A few grammar nitpicks, some questions, and a few suggestions.

I tested the following endpoints and was received an email or sms for each of them:

- ✅ `/notify/email`
- ✅ `/notify/sms`
- ✅ `/notify/email/attachment`
- ✅ `/notify/bulk/email`
- ✅ `/notify/bulk/sms`

esizer commented on api/app/Notify/Client.php:
```diff
@@ -0,0 +1,311 @@
+     *
+     * Send a single SMS with GC Notify
+     *
+     * @param string $to Phone number to send message to
```
I wish i could tell you but all they say in their docs is "The `phone_number` of the recipient of the text message."

mnigh replied:
Ok, I guess we are relying on their [underlying API](https://github.com/cds-snc/notification-api/blob/main/app/clients/sms/aws_sns.py#L28-L83) which itself uses AWS. Not sure we will ever send SMS, but if we do, hopefully we are storing a standard format that this works with. In any case, out of scope I think.

esizer commented on api/tests/Unit/NotifyTest.php:
```diff
@@ -0,0 +1,130 @@
+    private $username;
+    private $bulkName;
+
+    protected function setUp(): void
```
1. It actually is using a test API key (@tristan-orourke added it to our secrets for me)
2. We will need to update the API key off of my test account once we go to implement anyway
3. Also, we are technically doing smoke tests here and [you can use any API key](https://documentation.notification.canada.ca/en/testing.html#smoke-testing:~:text=You%20can%20use%20these%20smoke%20test%20numbers%20and%20addresses%20with%20any%20type%20of%20API%20key.) for them

esizer commented on api/config/notify.php:
```diff
@@ -0,0 +1,44 @@
+    | configuration
+    |
+    */
+    'templates' => [
```
Great idea, I moved them in [prep for merge](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/bcbd77e8a7a0e4f15501843cbe8d95762dbbe965)

esizer commented on api/routes/api.php:
```diff
@@ -18,3 +20,79 @@
 });
 
+/**
+ * Note: These routes are for testing/example and should be
```
Removed in [prep for merge](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/bcbd77e8a7a0e4f15501843cbe8d95762dbbe965)

esizer commented on .github/workflows/phpunit.yml:
```diff
@@ -10,6 +10,8 @@ jobs:
     container:
       image: kirschbaumdevelopment/laravel-test-runner:8.1
+      env:
+        GCNOTIFY_API_KEY: ${{ secrets.GCNOTIFY_API_KEY }}
```
Added in [prep for merge](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/bcbd77e8a7a0e4f15501843cbe8d95762dbbe965)

tristan-orourke commented:
Looking great! I was going to make some comments about bulk messages and moving config to env vars, but @mnigh already did 😁

mnigh commented on api/app/Exceptions/InvalidBulkRowDataException.php:
```diff
@@ -0,0 +1,18 @@
+namespace App\Exceptions;
+
+use Exception;
+
```
```suggestion
use Illuminate\Support\Facades\Log;
```

mnigh commented on api/app/Notify/Client.php:
```diff
@@ -93,24 +95,40 @@ public function sendSms($to, $template, $personalisation = [], $reference = null
-     * Send Bulk Notification
+     * Send Bulk Email Notification
+     *
+     * Send bulk Email with GC Notify
```
```suggestion
     * Send bulk email with GC Notify
```

mnigh commented on api/app/Notify/Client.php:
```diff
@@ -3,10 +3,12 @@
 namespace App\Notify;
 
 use Illuminate\Support\Facades\Http;
+use Illuminate\Support\Facades\Log;
```
Not referenced
```suggestion
```

esizer replied:
Removed in [removed unused import](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/11de08a3cbc11f6efd872feaaf5582bb4a9a62d5)

mnigh commented on api/app/Notify/Client.php:
```diff
@@ -226,6 +244,113 @@ private function buildBulkPayload($name, $rows, $template, $scheduleFor = null,
+    {
+
+        $normalizedRows = [['phone number']];
+        $personalisationKeyMap = null;
```
Not referenced anywhere, is this meant to be used?

esizer replied:
Removed in [cleanup + improve docs](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/2e8b6ca832c0d8ad2d923e837fece269b7f02663)

mnigh reviewed (changes requested):
I like the separation of concerns between bulk email and bulk sms. A few things to address though before I can approve.

esizer commented on api/app/Notify/Client.php:
```diff
@@ -226,6 +244,113 @@ private function buildBulkPayload($name, $rows, $template, $scheduleFor = null,
+
+        foreach($rows as $index => $row) {
+            if(!isset($row['email'])) {
+                throw new InvalidBulkRowDataException("Key 'email' not found in row $index.");
```
Updated in [collect bulk row data errors before throwing](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/2a8c03b13dc907a13b79a3f058dbc71a9486ac88)

esizer commented on api/app/Notify/Client.php:
```diff
@@ -226,6 +244,113 @@ private function buildBulkPayload($name, $rows, $template, $scheduleFor = null,
+
+        foreach($rows as $index => $row) {
+            if(!isset($row['phone_number'])) {
+                throw new InvalidBulkRowDataException("Key 'phone_number' not found in row $index.");
```
Updated in [collect bulk row data errors before throwing](https://github.com/GCTC-NTGC/gc-digital-talent/pull/5350/commits/2a8c03b13dc907a13b79a3f058dbc71a9486ac88)

mnigh reviewed (approved):
Great job! 🥇

mnigh commented on documentation/notifications.md:
```diff
@@ -0,0 +1,202 @@
+Sometimes, it is necessary to communicate with our users outside of the application. In order to do that, we can send emails and SMS messages using [GC Notify](https://notification.canada.ca/). This is done on the server in the `api` Laravel project.
+
+> **Note**
+> It may be useful to read over the [GC Notify documentation](https://documentation.notification.canada.ca/en/)
```
Allow docs site to choose language
```suggestion
> It may be useful to read over the [GC Notify documentation](https://documentation.notification.canada.ca/)
```

mnigh commented on documentation/notifications.md:
```diff
@@ -0,0 +1,202 @@
+
+### Using Templates
+
+You can access these Templates from the laravel config like any other configuration value.
```
```suggestion
You can access these templates from the Laravel config like any other configuration value.
```

mnigh commented on documentation/notifications.md:
```diff
@@ -0,0 +1,202 @@
+$emailTemplate = $templates['email'];
+
+// Getting a single template
+$emailTemplate = config('notify.templates.email')';
```
Remove orphan quote
```suggestion
$emailTemplate = config('notify.templates.email');
```

mnigh commented on documentation/notifications.md:
````diff
@@ -0,0 +1,202 @@
+);
+```
+
+### Bulk Emails
````
```suggestion
### Bulk Email
```

mnigh reviewed (changes requested):
A few minor suggestions, but otherwise, 🆒!

mnigh reviewed (approved):
🚀
